### ISSUE TITLE
Pretrained yolov5x6 gets 0.994 mAP@.5:.95 on coco valid split and 0.003 mAP@.5:.95 on test-dev split

### ISSUE THREAD

**Issue by Karol-G** [ultralytics/yolov5#5293]
Hi,

I used the pretrained yolov5x6 for evaluation on the official ms coco validation and test-dev split.
I cloned the code from the repo, no modifications have been done.

The following command was used for evaluation:

`python val.py --batch 28 --weights /home/k539i/Documents/experiments/yolov5/pretrained/yolov5x6.pt --data /home/k539i/Documents/datasets/original/mscoco2017/data.yaml --project /home/k539i/Documents/experiments/yolov5 --img 1280 --task val --device 0 --save-json --save-hybrid --augment`

`python val.py --batch 28 --weights /home/k539i/Documents/experiments/yolov5/pretrained/yolov5x6.pt --data /home/k539i/Documents/datasets/original/mscoco2017/data.yaml --project /home/k539i/Documents/experiments/yolov5 --img 1280 --task test --device 0 --save-json --save-hybrid --augment`

On the validation split I received a mAP@.5:.95 of 0.994 and on test-dev after submitting to the ms coco benchmark website I received a mAP@.5:.95 of 0.003. The generated predictions masks for the test-dev split look fine. I added some images for reference

My Questions:

- Did I do something obvious wrong?
- Has the pretrained model also been trained on the validation split?
- Why am I getting essentially 0 mAP on the test-dev split?
- Is it just an prediction output format mistake?

val log:
```
val: data=/home/k539i/Documents/datasets/original/mscoco2017/data.yaml, weights=['/home/k539i/Documents/experiments/yolov5/pretrained/yolov5x6.pt'], batch_size=28, imgsz=1280, conf_thres=0.001, iou_thres=0.6, task=val, device=0, single_cls=False, augment=True, verbose=False, save_txt=True, save_hybrid=True, save_conf=False, save_json=True, project=/home/k539i/Documents/experiments/yolov5, name=exp, exist_ok=False, half=False
YOLOv5 🚀 v6.0-4-gb754525e torch 1.9.1 CUDA:0 (NVIDIA GeForce RTX 3090, 24268.3125MB)

Fusing layers... 
Model Summary: 574 layers, 140730220 parameters, 0 gradients
val: Scanning '/home/k539i/Documents/datasets/original/mscoco2017/valid/labels.cache' images and labels... 5000 found, 0 missing, 48 empty, 0 corrupted: 100%|████████████████████████████████████████████████████████████████████████████████████████████████| 5000/5000 [00:00<?, ?it/s]
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 179/179 [08:39<00:00,  2.90s/it]
                 all       5000      36335          1      0.998      0.994      0.994
Speed: 0.2ms pre-process, 84.4ms inference, 1.9ms NMS per image at shape (28, 3, 1280, 1280)

Evaluating pycocotools mAP... saving /home/k539i/Documents/experiments/yolov5/exp17/yolov5x6_predictions.json...
loading annotations into memory...
pycocotools unable to run: [Errno 2] No such file or directory: '/home/k539i/Documents/datasets/original/mscoco2017/annotations/instances_val2017.json'
Results saved to /home/k539i/Documents/experiments/yolov5/exp17
5000 labels saved to /home/k539i/Documents/experiments/yolov5/exp17/labels
```

test-dev log:
```
val: data=/home/k539i/Documents/datasets/original/mscoco2017/data.yaml, weights=['/home/k539i/Documents/experiments/yolov5/pretrained/yolov5x6.pt'], batch_size=28, imgsz=1280, conf_thres=0.001, iou_thres=0.6, task=test, device=0, single_cls=False, augment=True, verbose=False, save_txt=True, save_hybrid=True, save_conf=False, save_json=True, project=/home/k539i/Documents/experiments/yolov5, name=exp, exist_ok=False, half=False
YOLOv5 🚀 v6.0-4-gb754525e torch 1.9.1 CUDA:0 (NVIDIA GeForce RTX 3090, 24268.3125MB)

Fusing layers... 
Model Summary: 574 layers, 140730220 parameters, 0 gradients
test: Scanning '/home/k539i/Documents/datasets/original/mscoco2017/test/images2.cache' images and labels... 0 found, 20288 missing, 0 empty, 0 corrupted: 100%|█████████████████████████████████████████████████████████████████████████████████████████████| 20288/20288 [00:00<?, ?it/s]
Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 725/725 [35:24<00:00,  2.93s/it]
                 all      20288          0          0          0          0          0
Speed: 0.2ms pre-process, 84.3ms inference, 1.7ms NMS per image at shape (28, 3, 1280, 1280)

Evaluating pycocotools mAP... saving /home/k539i/Documents/experiments/yolov5/exp16/yolov5x6_predictions.json...
loading annotations into memory...
pycocotools unable to run: [Errno 2] No such file or directory: '/home/k539i/Documents/datasets/original/mscoco2017/annotations/instances_val2017.json'
Results saved to /home/k539i/Documents/experiments/yolov5/exp16
20284 labels saved to /home/k539i/Documents/experiments/yolov5/exp16/labels
```

Results from ms coco benchmark website (https://competitions.codalab.org/competitions/20794):
```
overall performance
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.003
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.004
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.004
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.001
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.004
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.008
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.066
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.105
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.107
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.053
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.094
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.206
```

Test-dev split examples:
![val_batch0_labels](https://user-images.githubusercontent.com/3471895/138409359-3f4c31f3-2b48-4671-bcdc-f81d2ed1828d.jpg)

![val_batch0_pred](https://user-images.githubusercontent.com/3471895/138409373-3d1e449a-a183-478e-9d2e-eb2b60915ca9.jpg)

![val_batch1_labels](https://user-images.githubusercontent.com/3471895/138409380-a08874ba-90e3-45dd-9d54-8121c42bee31.jpg)

![val_batch1_pred](https://user-images.githubusercontent.com/3471895/138409383-fcbe84c1-119b-425e-92ca-039380d979cb.jpg)

![val_batch2_labels](https://user-images.githubusercontent.com/3471895/138409390-05172e08-2864-45e8-91e4-18c069119691.jpg)

![val_batch2_pred](https://user-images.githubusercontent.com/3471895/138409397-32aabab3-61d2-415d-9c60-62c19c3c38b2.jpg)

Best
Karol

**Comment by Karol-G** [comment]
I also evaluated with your given command in the pretrained section with:

`python val.py --data coco.yaml --img 1536 --iou 0.7 --augment`

Essentially the same result with a test-dev mAP@.5:.95 of 0.002

**Comment by glenn-jocher** [comment]
@Karol-G first, your commands to reproduce mAP are incorrect. The correct commands to reproduce are displayed in the README right under the results. To reproduce P6 results use `--img 1280`.

<img width="850" alt="Screenshot 2021-10-22 at 19 10 49" src="https://user-images.githubusercontent.com/26833433/138495936-c73b6d64-e85e-40ba-bdb0-0af374d197a3.png">

Second, we clearly do not have access to test-dev labels, so the expected mAP on test-dev locally should be zero. The entire purpose of the test-dev split is that the labels are not publicly available and that you must submit your predictions annotation JSON to the server as is stated in coco.yaml
https://github.com/ultralytics/yolov5/blob/15e8c4c15bff0e3c729868e2c31f80f918e31fb2/data/coco.yaml#L14

**Comment by Karol-G** [comment]
I used the command under the point TTA for yolov5x6 as the pretrained model was trained with image size of 1536. So I should still use your highlighted command?

I reported the mAP results for the test-dev set from the coco benchmark website (https://competitions.codalab.org/competitions/20794) and showed their score log in the post.

Here again:

```
overall performance
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.003
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.004
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.004
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.001
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.004
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.008
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.066
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.105
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.107
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.053
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.094
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.206
```

But if one only gets good results with your highlighted command for the yolov5x6 pretrained model than I will reevaluate the test-dev set again.

**Comment by glenn-jocher** [comment]
Results on val are simple to reproduce:

<img width="1243" alt="Screenshot 2021-10-22 at 19 23 01" src="https://user-images.githubusercontent.com/26833433/138497308-e878c11b-d9e3-4e6d-b296-8c76909306c9.png">

**Comment by glenn-jocher** [comment]
@Karol-G I've submitted YOLOv5x6 to the COCO server for test eval, result is 55.8 mAP@0.5:0.95, a little bit higher than reported on val set. Results submitted to leaderboard: https://competitions.codalab.org/competitions/20794#results

```bash
overall performance
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.558
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.727
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.615
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.388
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.594
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.679
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.406
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.678
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.728
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.561
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.764
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.861
```

Anyone can do the same though. To reproduce run val command below, zip `runs/val/exp/yolov5x6_predictions.json`, and submit to [server](https://competitions.codalab.org/competitions/20794#participate) under Participate tab.
```
!python val.py --weights yolov5x6.pt --data coco.yaml --img 1536 --iou 0.70 --half --task test --augment
```

**Comment by Karol-G** [comment]
Thanks! I will try it on monday. But really thanks for the great support and patience with me :)

Best
Karol

**Comment by Karol-G** [comment]
I was finally able to reproduce the test-dev results locally and on colab.

The reason I got bad results locally was probably due to the some wrong preparation of the test-dev dataset. After I redownloaded it again everything worked fine and I got the same results as you.

On colab the issue was a different one:
When executing the cell to download the test-dev set in the section "COCO test", the test-dev set is downloaded to "/content/datasets/coco" as the cell points to "../datasets" and not to "./datasets" (same for val cell). When I then execute the validation cell the used coco.yaml does not point to the just downloaded test-set but to "/content/yolov5/datasets/coco". As a consequence, YOLO tries to download the entire coco dataset when running val.py. Eventually, the colab notebook will run out of space, but will continue to execute the cell. So most of the images can not be read and thus get skipped during validation, which again results in the bad eval score when submitting to the website.

Fix would be to point to "./datasets" instead of ../datasets

Best
Karol

**Comment by glenn-jocher** [comment]
@Karol-G interesting. The assumption is that the commands are all run from yolov5/, if you run them elsewhere then yes this may cause problems.

**Comment by Karol-G** [comment]
I ran them actually from within the yolov5 dir

**Comment by glenn-jocher** [comment]
@Karol-G were you running the same command that's in Colab? I fixed this a few days ago:
```python
# Download COCO test-dev2017
torch.hub.download_url_to_file('https://ultralytics.com/assets/coco2017labels.zip', 'tmp.zip')
!unzip -q tmp.zip -d ../datasets && rm tmp.zip
!f="test2017.zip" && curl http://images.cocodataset.org/zips/$f -o $f && unzip -q $f -d ../datasets/coco/images
```

**Comment by Karol-G** [comment]
yes

**Comment by Karol-G** [comment]
It should be reproducable with the newest notebook

**Comment by glenn-jocher** [comment]
Oh strange. I used the newest notebook to produce my results. The correct layout is shown in coco.yaml:
```bash
# content
# ├── yolov5
# └── datasets
#     └── coco  ← downloads here
```

**Comment by Karol-G** [comment]
I'll try to reproduce it again, one minute.

**Comment by Karol-G** [comment]
Well, nvm. It is fixed in the newest version

**Comment by glenn-jocher** [comment]
@Karol-G oh great! :)

**Comment by HLH13297997663** [comment]
Hi:
   You received a mAP@.5:.95 of 0.994 in validation dataset, Why is mAP so great？I  received a mAP@.5:.95 of 0.524 after 24 epoch training. Did I make some mistake in traning?

**Comment by HLH13297997663** [comment]
@Karol-G

**Comment by PraveenMNaik** [comment]
@glenn-jocher @Karol-G , i am experimenting with new yaml file of Ghosgtnet, howevr I could not be able to get result for coco dataset , given 0 for map, precison, recall throughout... pls suggest solution

**Comment by glenn-jocher** [comment]
@PraveenMNaik hello,

There could be several reasons for getting 0 mAP, precision, recall on coco dataset:

1. Insufficient training data: Training YOLOv5 on coco dataset requires a large amount of training data, preferably covering all object classes. If you are only training on a small subset of coco dataset, then this could result in poor performance.

2. Overfitting: If you are seeing good performance on your training data but poor performance on the validation or test dataset, then your model may be overfitting. This can be addressed by using regularization techniques such as dropout and weight decay.

3. Incorrect configuration: You may want to cross-check your configuration files such as the data.yaml, train.yaml and .yml hyperparameter files as one wrong configuration parameter can result in poor performance.

4. Low-quality training images: It is generally recommended to use high-quality training images of a consistent size and resolution to train YOLOv5. Using low-quality images with varying resolutions can lead to poor performance.

5. Improper annotations: improper annotations such as incomplete bounding boxes, incorrect bounding boxes or incorrect class labels can impact the performance of the model.

Please check if any of these reasons apply to your case and take necessary corrective actions accordingly.

Thank you
